libpairinteraction/MatrixElementCache.cpp

```diff
@@ -0,0 +1,763 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Communication.h"
+#include "MatrixElementCache.h"
+#include "QuantumDefect.h"
+#include "SQLite.h"
+#include "version.h"
+
+#include <sstream>
+#include <iostream>
+#include <string>
+#include <limits>
+#include <stdexcept>
+#include <cctype>
+#include <memory>
+
+bool selectionRulesMomentumNew(StateOne const& state1, StateOne const& state2, int q) {
+    bool validL = state1.l == state2.l;
+    bool validJ = fabs(state1.j-state2.j) <= 1;
+    bool validM = state1.m == state2.m+q;
+    bool validQ = abs(q) <= 1;
+    return validL && validJ && validM && validQ;
+}
+
+bool selectionRulesMomentumNew(StateOne const& state1, StateOne const& state2) {
+    bool validL = state1.l == state2.l;
+    bool validJ = fabs(state1.j-state2.j) <= 1;
+    bool validM = (fabs(state1.m - state2.m) <= 1);
+    return validL && validJ && validM;
+}
+
+bool selectionRulesMultipoleNew(StateOne const& state1, StateOne const& state2, int kappa, int q) {
+    bool validL = (abs(state1.l-state2.l) <= kappa) && (kappa%2 == abs(state1.l-state2.l)%2);
+    bool validJ = (fabs(state1.j-state2.j) <= kappa)  && (state1.j+state2.j >= kappa);
+    bool validM = state1.m == state2.m+q;
+    bool validQ = abs(q) <= kappa;
+    bool noZero = !(kappa == 2 && state1.j == state2.j && state2.j == 1.5 && state1.m == -state2.m && fabs(state1.m - state2.m) == 1);
+    return validL && validJ && validM && validQ && noZero;
+}
+
+bool selectionRulesMultipoleNew(StateOne const& state1, StateOne const& state2, int kappa) {
+    bool validL = (abs(state1.l-state2.l) <= kappa) && (kappa%2 == abs(state1.l-state2.l)%2);
+    bool validJ = (fabs(state1.j-state2.j) <= kappa)  && (state1.j+state2.j >= kappa);
+    bool validM = (fabs(state1.m - state2.m) <= kappa);
+    bool noZero = !(kappa == 2 && state1.j == state2.j && state2.j == 1.5 && state1.m == -state2.m && fabs(state1.m - state2.m) == 1);
+    return validL && validJ && validM && noZero;
+}
+
+////////////////////////////////////////////////////////////////////
+/// Constructors ///////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////
+
+MatrixElementCache::MatrixElementCache() : defectdbname(""), dbname(""), db(dbname), stmt(db), pid_which_created_db(utils::get_pid()) { // db and stmt have to be initialized here since they have no default constructor
+    method = "Modelpotentials";
+    //if (config["missingCalc"].str() == "true") {
+    //    method = "Modelpotentials";
+    //} else if (config["missingWhittaker"].str() == "true") {
+    //    method = "Whittaker";
+    //} else {
+    //    method = "Error";
+    //}
```

Comment:
Is it intended that the method switching for the radial matrix elements is commented out? (same below)